### ISSUE TITLE
Informative message in case of having explored the entire experimental space

### ISSUE THREAD

**Issue by DitlevHB** [ProtoLife/daptics-api#6]
When optimizing a simple experiment you would eventually reach the point of having explored the entire experimental space and there would be no more generations to iterate. 
Currently an exception is thrown with the message ("no task found").
Perhaps a better, more informative message, could be in the response?

**Comment by ggazzola** [comment]
You are right. The web interface is already equipped to deal with this situation... Within a session, the sidebar on the left contains an item called "Remaining generations". Its value incrementally decreases every time a generation is run (also keeping into account extra experiments), reaching zero as soon as there are fewer experiments left than the population size. At that point, the web interface displays a message and prevents the user from proceeding further. In the API, we could use a similar approach that relies on the same "Remaining generations" counter.

**Comment by pzingg** [comment]
I am working on the next release of the API, which will expose "remaining" and "completed" values for the session. I have a couple of questions:

1. Is there something in the API we should offer the user when "remaining" goes to zero?  Didn't we have some final "campaign completion" experiments upload or something in the web application when there were no more remaining generations?

2. Should the "completed" flag be set only AFTER the final "campaign completion" upload is done?  In which case we have four states:

a. "remaining" is unknown and "completed is false (before the ESD is saved?)
b. "remaining" is a positive number and "completed" is false - meaning backend will still generate designs
c. "remaining" is zero  and "completed" is false - meaning backend will refuse to generate designs, but user can upload "final experiments".
d. "remaining" is zero and "completed" is true - meaning no uploads or designs to be done.

**Comment by pzingg** [comment]
In the "dev" version ot the API I am testing with now, when a user tries to generate a design without paying attention to the "remaining" number, the API will attempt to start a design generation task, but the task will fail quickly with this error message:

``
numAvailableExperiments >= number is not TRUE
```
I will make some changes that will prevent the task from being attempted on the backend if "remaining" is zero, as long as that value is correct, and give the user a message like "No generations are remaining", or whatever you think would be a helpful message.

**Comment by pzingg** [comment]
Question: Can you "finalize" a campaign even when remaining is greater than zero?  You don't want to pursue optimization any further, just record the results of the last design?

**Comment by ggazzola** [comment]
> Is there something in the API we should offer the user when "remaining" goes to zero? Didn't we have some final "campaign completion" experiments upload or something in the web application when there were no more remaining generations?

Yes, in the web app, this is what happens when "Remaining generations" reaches 0:

1) The experiments view displays this (note the red-highlighted message + new 'Finish Campaign' button)

<img width="895" alt="Screen Shot 2019-05-09 at 6 19 42 PM" src="https://user-images.githubusercontent.com/11232211/57491062-614c7f00-7289-11e9-93bf-6b99ffd32e2e.png">

If then the user chooses to specify responses for the current (last) generation (plus possibly extra experiments + responses), we then get the backend to:

2) Incorporate extra experiments (cfr. `AutoIncorporateExperiments.sh`), if any

3) Update responses (cfr. `AutoUpdateCurrentGen.sh`)

4) Take a snapshot (cfr. `AutoSnapshot.sh`)

(Note no call to`AutoSampleNextGen.sh` etc.).

5) Finally, we display 

<img width="791" alt="Screen Shot 2019-05-09 at 6 20 31 PM" src="https://user-images.githubusercontent.com/11232211/57491174-c30ce900-7289-11e9-81b9-4de34b6fd3f5.png">

which replaces the usual HOT, and therefore doesn't allow the user to proceed with any further generations.

I think in the API we could display a warning message similar to that in (1) and then, if the user wants to proceed, do (2)–(4), and finally display an error message if the user tries to produce another generation.

>So the "completed" flag be set only AFTER the final response upload is done. In which case we have four states:

>a. "remaining" is unknown and "completed is false (before the ESD is saved?)
>b. "remaining" is a positive number and "completed" is false - meaning backend will still generate designs
>c. "remaining" is zero and "completed" is false - meaning user should upload final experiment responses but backend will NOT generate another design.
>d. "remaining" is zero and "completed" is true - meaning no uploads or designs will be done.

Yes, that seems correct.

> Question: Can you "finalize" a campaign even when remaining is greater than zero? You don't want to pursue optimization any further, just record the results of the last design?

Backend-wise, yes (you can just skip `AutoSampleNextGen.sh` etc), but currently we don't have this option in the web app.

**Comment by pzingg** [comment]
See PR #7. Version 0.7.0 of the backend and API has support for getting the "remaining" and "completed" status of a session, and also adds the `finalize_campaign` method.

When the backend server has been upgraded to v0.7.0, we will merge PR #7 here.
